### ISSUE TITLE
使用 ms-swift 对 Qwen2.5-VL-3B-Instruct 做 GUI-R1 风格 SFT 后，坐标点击任务推理准确率极低

### ISSUE THREAD

**Issue by zhonganbi102-netizen** [modelscope/ms-swift#6900]
您好，我在使用 ms-swift 对 `Qwen/Qwen2.5-VL-3B-Instruct` 做 GUI-R1 风格的全参数 SFT（点击坐标任务）时遇到一个问题：  
**训练 loss / token_acc 看起来正常收敛，但在下游 Screenspot 点击任务上的实际点击坐标准确率非常低。**  
想请教下这是否和我使用的数据格式 / 绝对坐标标签 / 配置有关，还是我哪里用 ms-swift 的方式不对。

---

### 1. 环境与模型

- 框架：`ms-swift`（当前 master，对应 `swift/cli/sft.py`）
- 基础模型：`Qwen/Qwen2.5-VL-3B-Instruct`
- GPU：单机 3 卡 A100 40GB（CUDA_VISIBLE_DEVICES=1,2,4）
- 训练类型：`train_type=full`（全参数 SFT）

---

### 2. 训练脚本（ms-swift）

训练脚本为 `scripts/train/baseline.sh`，内容如下

```bash
nnodes=1
nproc_per_node=3
export ROOT_IMAGE_DIR="/home/users/ju17777/manbing/bizhongan2/linxid/data/GUI-R1-train"
export MAX_PIXELS=1048576

TRAIN_JSONL="/home/users/ju17777/manbing/bizhongan2/linxid/data/GUI-R1-train/train.jsonl"
VAL_JSONL="/home/users/ju17777/manbing/bizhongan2/linxid/data/GUI-R1-train/test.jsonl"

CUDA_VISIBLE_DEVICES=1,2,4 \
torchrun \
    --master_port 29500 \
    --nproc_per_node=$nproc_per_node \
    --nnodes=$nnodes \
    --node_rank=0 \
    --master_addr=127.0.0.1 \
    swift/cli/sft.py \
    --model Qwen/Qwen2.5-VL-3B-Instruct \
    --dataset "${TRAIN_JSONL}" \
    --val_dataset "${VAL_JSONL}" \
    --train_type full \
    --torch_dtype bfloat16 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --learning_rate 1e-5 \
    --freeze_vit false \
    --freeze_llm false \
    --freeze_aligner false \
    --gradient_accumulation_steps 1 \
    --save_steps 1000 \
    --save_total_limit 2 \
    --logging_steps 5 \
    --max_length 8192 \
    --output_dir output \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 3
```

训练结果（`logging.jsonl`）里最后几行摘要大致为：

- `train_loss ≈ 0.306`
- `eval_loss  ≈ 0.265`
- `eval_token_acc ≈ 0.90`

看起来 SFT 是正常收敛的。

---

### 3. 训练/验证数据格式（绝对坐标标签）

我自己构造了一份 GUI-R1 风格的数据，放在：

- 训练集：`.../GUI-R1-train/train.jsonl`
- 验证集：`.../GUI-R1-train/test.jsonl`

每一行是一个 JSON，对应 `ms-swift` 所需的多轮对话格式，示例（train / val 都类似）：

```json
{
  "messages": [
    {
      "role": "user",
      "content": "<image>\nYou are GUI-R1, a reasoning GUI Agent Assistant. In this UI screenshot, I want you to continue executing the command 'click the UI element plateforme', with the action history being 'None'.\nPlease provide the action to perform (enumerate from ['click']), the point where the cursor is moved to (integer) if a click is performed, and any input text required to complete the action.\nOutput the thinking process in <think> </think> tags and the final answer in <answer> </answer> tags as follows:\n<think> ... </think> <answer>[{'action': enum[ 'click'], 'point': [x, y], 'input_text': 'no input text'}]</answer>\nExample:\n[{'action': enum['click'], 'point': [123, 300], 'input_text': 'no input text'}]\n"
    },
    {
      "role": "assistant",
      "content": "<think></think><answer>[{\"action\": \"enum[click]\", \"point\": [1004, 655], \"input_text\": \"no input text\"}]</answer>"
    }
  ],
  "images": ["images/train/000000.jpg"]
}
```

**几点说明：**

- `images` 字段存的是相对路径（`ROOT_IMAGE_DIR` 下的图片），我理解 ms-swift 会自动根据 `--dataset` 和 `images` 字段去读图。
- 这里的 `point: [x, y]` 是 **屏幕的绝对像素坐标**，没有做归一化或任何缩放处理。
- `<think>` 标签目前都是空的（训练时没有真正的思维过程内容），只监督 `<answer>` 里的 JSON。

---

### 4. 推理与评测方式

推理和评测是通过我自己写的脚本完成的，**不在 ms-swift 里面**，但我想确认 ms-swift 训练出来的模型是否有我没考虑到的注意事项。

- 推理使用 vLLM，对 `Qwen2.5-VL-3B-Instruct` 和我 SFT 后的 checkpoint 都做了统一的 inference：
  - 模型路径（SFT 后）：  
    `/home/users/ju17777/manbing/bizhongan2/linxid/ms-swift-main/output/v10-20251203-210341/checkpoint-1190`
  - 评测数据：公开的 Screenspot 测试集  
    `/home/users/ju17777/manbing/bizhongan2/demand1-bizhongan/verl/datasets/GUI-R1-raw/screenspot_test.parquet`
  - 该 parquet 里每条样本有字段：`image`（bytes）、`instruction`、`gt_bbox`、`group`、`ui_type`。

在推理脚本 `inference_vllm_screenspot.py` 里，我对每条 `instruction` 拼接了和 SFT 时几乎完全一致的 prompt，然后用 Qwen 官方的 `AutoProcessor` 生成输入：

```python
text = sample["instruction"]
text = (
    f"You are GUI-R1, a reasoning GUI Agent Assistant. In this UI screenshot, I want you to continue executing the command '{text}', with the action history being 'None'.\n"
    "Please provide the action to perform (enumerate from ['click']), the point where the cursor is moved to (integer) if a click is performed, and any input text required to complete the action.\n"
    "Output the thinking process in <think> </think> tags, and the final answer in <answer> </answer> tags as follows:\n"
    "<think> ... </think> <answer>[{'action': enum[ 'click'], 'point': [x, y], 'input_text': 'no input text'}]</answer>\n"
    "Example:\n"
    "[{'action': enum['click'], 'point': [123, 300], 'input_text': 'no input text'}]\n"
)
text = "<image>\n" + text
```

模型输出示例（SFT 后模型）：

```json
{
  "image": "",
  "instruction": "view the language & region settings",
  "gt_bbox": [1018, 368, 1139, 515],
  "group": "desktop",
  "ui_type": "text",
  "image_size": [1344, 1372],
  "pred": "<think></think><answer>[{\"action\": \"enum[click]\", \"point\": [1647, 658], \"input_text\": \"no input text\"}]</answer>",
  "pred_coord": [1647.0, 658.0]
}
```

可以看到：

- 输出格式和训练时完全一致：`<think></think><answer>[{...}]</answer>`；
- `<think>` 也是空的，这是因为训练数据里从未提供过非空 `<think>`，这一点我认为是合理的。

---

### 5. 问题：坐标点击准确率极低

在 `screenspot_test.parquet` 全量上，我用一个简单的评估脚本（按照公式判断点击点是否落在 `gt_bbox` 内、或 IoU/距离阈值）做了统计，结果非常差：

```text
Type desktop-text : 0.1031
Type desktop-icon : 0.0786
Type mobile-text  : 0.0440
Type mobile-icon  : 0.0524
Type web-text     : 0.0565
Type web-icon     : 0.0340
```

相比之下，**同样的评测方法**下，base 模型（未用 ms-swift SFT）在同一数据上的表现明显更好。  
也就是说：**全参数 SFT 之后，在这个点击坐标任务上性能严重退化**。

我目前的理解是：

- 这可能和 **我用的 SFT 数据分布**（只包含某一类 GUI-R1 风格 click 任务）、以及使用“绝对坐标标签”有关；
- 也有可能是我对 `--dataset`（JSONL with messages + images 路径）的使用方式存在问题，比如 image 读取、max_pixels、坐标缩放之类的细节，导致模型在训练/推理时看到的图像坐标系不一致。

---

### 6. 想请教的问题

1. **关于数据格式 / 坐标标签**

   - 像这种 GUI 点击坐标任务，直接在 `assistant` 的文本里用**屏幕的绝对坐标 `[x, y]`** 做监督（不做归一化），对 ms-swift 来说是否 OK？  
   - 在 Vision-Language SFT 场景下，`ROOT_IMAGE_DIR` + `images` 相对路径的这种用法，是否还需要在参数里额外指定某些字段（例如图像分辨率 / max_pixels），避免训练时和我后面 vLLM 推理时的 `AutoProcessor` 行为不一致？

2. **关于训练配置**

   - 我目前是 `train_type=full` 做全参数 SFT，`num_train_epochs=1`，`lr=1e-5`。  
     在你们的经验里，这种“数据量不大 + 分布相对集中”的情况下，是否更推荐：
     - 使用 LoRA / 部分参数 SFT，而不是 full；  
     - 或者减小学习率、增加 epoch，以避免破坏原有的视觉/坐标感知能力？
   - `eval_token_acc` 在 0.9 左右，这个指标对这种“输出 JSON + 坐标”的任务是否有参考意义？还是更适合 NLP token 任务？

3. **是否有推荐的做法 / 示例**

   - ms-swift 里有没有 **Vision-Language 坐标/检测类任务** 的官方示例？  
     例如：怎么推荐组织 JSONL 数据、是否需要对坐标做归一化或 embedding 处理，而不是直接塞到文本中。
   - 如果有类似 GUI / Screenspot 这类 benchmark 的推荐实践（特别是如何在不破坏原有能力的同时做 SFT），也非常希望能参考。

---
